operator/pkg/ciliumendpointslice/manager.go

```diff
@@ -368,6 +371,110 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
 	return
 }
 
+// CompressCESsOnFCFSStartUp makes sure that once FCFS (First Come, First
+// Served) CES manager starts, all CESs will be compressed to a minimum number
+// of fully packed CESs.
+//
+// This is only required when changing the batching strategy from Identity to
+// FCFS, to ensure expected performance. Otherwise, it will perform no action,
+// as all CESs are already compressed by design with FCFS batching.
+func (c *cesMgr) CompressCESsOnFCFSStartUp() {
+	log.Info("Running CompressCESsOnFCFSStartUp.")
+	// The compression needs to be done per namespace, since one CES cannot
+	// contain CEPs with different namespaces.
+	cesPerNamespace := make(map[string][]*cesTracker)
+	for _, ces := range c.desiredCESs.desiredCESs {
+		cesPerNamespace[ces.ces.Namespace] = append(cesPerNamespace[ces.ces.Namespace], ces)
+	}
+
+	for namespace, cesList := range cesPerNamespace {
+		c.compressCESList(cesList, namespace)
+	}
+
+	log.Info("CompressCESsOnFCFSStartUp completed.")
+}
+
+func (c *cesMgr) compressCESList(cesList []*cesTracker, namespace string) {
+	if len(cesList) == 0 {
+		log.WithFields(logrus.Fields{
+			logfields.K8sNamespace: namespace,
+		}).Info("No action required for compressCESList. No CES currently exist.")
+		return
+	}
+
+	totalCEPs := 0
+	for _, ces := range cesList {
+		ces.backendMutex.RLock()
```

Comment:
This code either doesn't need to be locked or you will have problems.

Why do you lock to get number of enpoints?

Reply:
It's the same method used in the getTotalCEPCount() for the cached CESs.
https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/manager.go#L372

I use a separate list here, but they still point to existing CESs (cesTrackers) that could be changing, and the lock is used to avoid incorrect read during the change.